### PR TITLE
feat: support obtaining license data from secret

### PR DESCRIPTION
**Pull request by Gala-R**

### CODE REVIEW

File: apis/apps/v1beta3/webhook_validation_test.go [hunk 1]
```diff
@@ -179,3 +227,29 @@ func checkValidation(emqx Emqx) {
 	obj.SetUsername("test")
 	Expect(k8sClient.Update(context.Background(), obj)).ShouldNot(Succeed())
 }
+
+func checkLicenseValidation(emqx *EmqxEnterprise) {
```
Comment (Rory-Z) [inline apis/apps/v1beta3/webhook_validation_test.go:231]: Please add more test case about secretName

File: controllers/apps/v1beta3/emqx_handler.go [hunk 2]
```diff
@@ -128,22 +129,27 @@ func (r *EmqxReconciler) Do(ctx context.Context, instance appsv1beta3.Emqx) (ctr
...
 	}
 
-	license := generateLicense(instance)
-	if license != nil {
-		resources = append(resources, license)
-		sts = updateLicenseForsts(sts, license)
+	emqxEnterprise, ok := instance.(*appsv1beta3.EmqxEnterprise)
+	if ok {
+		license := &corev1.Secret{}
+		err = r.Client.Get(context.Background(), types.NamespacedName{Name: emqxEnterprise.GetLicense().SecretName, Namespace: emqxEnterprise.GetNamespace()}, license)
```
Comment (Rory-Z) [inline controllers/apps/v1beta3/emqx_handler.go:145]: First check secretName not equal ""

File: controllers/apps/v1beta3/emqx_handler.go [hunk 3]
```diff
@@ -128,22 +129,27 @@ func (r *EmqxReconciler) Do(ctx context.Context, instance appsv1beta3.Emqx) (ctr
...
+		if err != nil || len(license.Data) != 1 { //only one key-value otherwise is invalid
```
Comment (Rory-Z) [inline controllers/apps/v1beta3/emqx_handler.go:146]: No need to check `license.Data`, because 620 line have do this

**Review by Rory-Z — COMMENTED** [review]
Why are more than 8000 changes  to `deploy/manifests/emqx-operator-controller.yaml` ?

File: controllers/apps/v1beta3/emqx_handler.go [hunk 4]
```diff
@@ -970,10 +976,16 @@ func updateLoadedModulesForSts(sts *appsv1.StatefulSet, loadedModules *corev1.Co
...
+	for k := range license.Data {
+		fileName = k
+		break
+	}
+
 	return updateEnvAndVolumeForSts(sts,
 		corev1.EnvVar{
 			Name:  "EMQX_LICENSE__FILE",
-			Value: "/mounted/license/emqx.lic",
+			Value: fmt.Sprintf("%s/%s", "/mounted/license", fileName),
```
Comment (Rory-Z) [inline controllers/apps/v1beta3/emqx_handler.go:988]: Can use  `filepath:join`

File: controllers/apps/v1beta3/emqx_handler.go [hunk 5]
```diff
@@ -128,22 +129,31 @@ func (r *EmqxReconciler) Do(ctx context.Context, instance appsv1beta3.Emqx) (ctr
...
 		resources = append(resources, loadedModules)
 		sts = updateLoadedModulesForSts(sts, loadedModules)
 	}
 
-	license := generateLicense(instance)
-	if license != nil {
-		resources = append(resources, license)
-		sts = updateLicenseForsts(sts, license)
+	emqxEnterprise, ok := instance.(*appsv1beta3.EmqxEnterprise)
+	if ok {
```
Comment (Rory-Z) [inline controllers/apps/v1beta3/emqx_handler.go:143]:
```suggestion
	if emqxEnterprise, ok := instance.(*appsv1beta3.EmqxEnterprise); ok {
```

File: controllers/apps/v1beta3/emqx_handler.go [hunk 6]
```diff
@@ -128,22 +130,30 @@ func (r *EmqxReconciler) Do(ctx context.Context, instance appsv1beta3.Emqx) (ctr
...
-	license := generateLicense(instance)
-	if license != nil {
-		resources = append(resources, license)
-		sts = updateLicenseForsts(sts, license)
+	if emqxEnterprise, ok := instance.(*appsv1beta3.EmqxEnterprise); ok {
+		var license *corev1.Secret
+		if emqxEnterprise.GetLicense().SecretName != "" {
+			license = &corev1.Secret{}
+			err := r.Client.Get(context.Background(), types.NamespacedName{Name: emqxEnterprise.GetLicense().SecretName, Namespace: emqxEnterprise.GetNamespace()}, license)
+			return ctrl.Result{RequeueAfter: time.Duration(5) * time.Second}, err
```
Comment (Rory-Z) [inline controllers/apps/v1beta3/emqx_handler.go:148]:
```suggestion
			if err := r.Client.Get(context.Background(), types.NamespacedName{Name: emqxEnterprise.GetLicense().SecretName, Namespace: emqxEnterprise.GetNamespace()}, license); err != nil {
			    return ctrl.Result{}, err
			}
			
```

File: controllers/apps/v1beta3/emqx_handler.go [hunk 7]
```diff
@@ -128,22 +130,30 @@ func (r *EmqxReconciler) Do(ctx context.Context, instance appsv1beta3.Emqx) (ctr
...
-	license := generateLicense(instance)
-	if license != nil {
-		resources = append(resources, license)
-		sts = updateLicenseForsts(sts, license)
+	if emqxEnterprise, ok := instance.(*appsv1beta3.EmqxEnterprise); ok {
+		var license *corev1.Secret
+		if emqxEnterprise.GetLicense().SecretName != "" {
+			license = &corev1.Secret{}
+			err := r.Client.Get(context.Background(), types.NamespacedName{Name: emqxEnterprise.GetLicense().SecretName, Namespace: emqxEnterprise.GetNamespace()}, license)
+			return ctrl.Result{}, err
```
Comment (Rory-Z) [inline controllers/apps/v1beta3/emqx_handler.go:148]: if err != nil {}

File: apis/apps/v1beta3/emqxenterprise_types.go [hunk 8]
```diff
@@ -22,8 +22,12 @@ import (
 )
 
 type License struct {
-	Data       []byte `json:"data,omitempty"`
+	// Data represents the byte data of the license content
```
Comment (Rory-Z) [inline apis/apps/v1beta3/emqxenterprise_types.go:25]: Data contains the secret data. Each key must consist of alphanumeric characters, '-', '_' or '.'. The serialized form of the secret data is a base64 encoded string, representing the arbitrary (possibly non-string) data value here. Described in https://tools.ietf.org/html/rfc4648#section-4

File: apis/apps/v1beta3/emqxenterprise_types.go [hunk 9]
```diff
@@ -22,8 +22,12 @@ import (
...
+	Data []byte `json:"data,omitempty"`
+	// StringData represents the string data of the license content
```
Comment (Rory-Z) [inline apis/apps/v1beta3/emqxenterprise_types.go:27]: stringData allows specifying non-binary secret data in string form. It is provided as a write-only input field for convenience. All keys and values are merged into the data field on write, overwriting any existing values.

File: apis/apps/v1beta3/emqxenterprise_types.go [hunk 10]
```diff
@@ -22,8 +22,12 @@ import (
...
 	StringData string `json:"stringData,omitempty"`
+	// SecretName represents the license content from the specified secret
```
Comment (Rory-Z) [inline apis/apps/v1beta3/emqxenterprise_types.go:29]: secretName is the name of the secret in the pod's namespace to use. More info: https://kubernetes.io/docs/concepts/storage/volumes#secret

File: controllers/apps/v1beta3/emqx_handler.go (hunk 2 above)
Comment (Gala-R) [reply controllers/apps/v1beta3/emqx_handler.go]: test

**Review by Rory-Z — APPROVED** [review]

**Pull request merged**